examples/DataApi-Water-And-Energy-Usage/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -12,69 +12,218 @@ import org.http4s.HttpRoutes
-    IO {
-      println(oldState)
-      updates
-        .map(_.value)
-        .map {
-          case update: EnergyUsage =>
-            if (update.value > 0L) ().validNec else EnergyNotPositive.asInstanceOf[DataApplicationValidationError].invalidNec
-          case update: WaterUsage =>
-            if (update.value > 0L) ().validNec else WaterNotPositive.asInstanceOf[DataApplicationValidationError].invalidNec
+  @derive(decoder, encoder)
+  case class Update(address: Address, energyUsage: Option[EnergyUsage], waterUsage: Option[WaterUsage]) extends DataUpdate
+
+  @derive(decoder, encoder)
+  case class State(devices: Map[Address, AggregatedUsage]) extends DataState
+
+  def validateData(oldState: State, updates: NonEmptyList[Signed[Update]])(implicit sp: SecurityProvider[IO]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateUpdates = updates.traverse { update =>
+      validateUpdate(update.value)
+    }.map(_.reduce)
+
+    val validateAddress = updates.traverse { update =>
+        val p = update.proofs
+          .map(_.id)
+          .toList
+          .traverse(_.toAddress[IO])
+          .map(validateProvidedAddress(_, update.value.address))
+        p
+      }.map(_.reduce)
+
+
+    val validateTimestamp = IO {
+      updates.map { signedUpdate =>
+        val update = signedUpdate.value
+        val energyUsage = update.energyUsage.getOrElse(EnergyUsage.empty)
+        val waterUsage = update.waterUsage.getOrElse(WaterUsage.empty)
+        val address = update.address
+
+        if (energyUsage.timestamp > 0 && waterUsage.timestamp > 0) {
+          validateEnergyTimestampToNoDoubling(oldState.devices.get(address), energyUsage).productR(validateWaterTimestampToNoDoubling(oldState.devices.get(address), waterUsage))
+        } else if (energyUsage.timestamp > 0) {
+          validateEnergyTimestampToNoDoubling(oldState.devices.get(address), energyUsage)
+        } else if (waterUsage.timestamp > 0) {
+          validateWaterTimestampToNoDoubling(oldState.devices.get(address), waterUsage)
+        } else {
+          ().validNec
         }
-        .reduce
+      }.reduce
     }
 
-  def validateUpdate(update: Usage): IO[DataApplicationValidationErrorOr[Unit]] = IO {
-    update match {
-      case u: EnergyUsage =>
-        if (u.value > 0L) ().validNec else EnergyNotPositive.asInstanceOf[DataApplicationValidationError].invalidNec
-      case u: WaterUsage =>
-        if (u.value > 0L) ().validNec else WaterNotPositive.asInstanceOf[DataApplicationValidationError].invalidNec
-    }
+    for {
+      validatedUpdates <- validateUpdates
+      validatedAddress <- validateAddress
+      validatedTimestamp <- validateTimestamp
+    } yield validatedUpdates.productR(validatedAddress).productR(validatedTimestamp)
+
   }
 
-  def combine(oldState: State, updates: NonEmptyList[Signed[Usage]]): IO[State] =
+  def validateUpdate(update: Update): IO[DataApplicationValidationErrorOr[Unit]] = {
```

Comment:
I think we should have descriptive comments for each of these major methods (validateData, validateUpdate, combine, serialize*) describing what their role is and what we're doing in the code. This is an example app so it would be helpful to have much more verbose comments than usual.